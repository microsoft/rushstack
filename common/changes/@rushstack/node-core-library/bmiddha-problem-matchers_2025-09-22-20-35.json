{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `FileError.getProblemMatcher()` which returns the problem matcher compatible with `IOperationExecutionResult.problemCollector` as well as VS Code, GitHub Actions",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}
