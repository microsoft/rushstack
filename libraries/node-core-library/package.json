{
  "name": "@rushstack/node-core-library",
  "version": "5.10.1",
  "description": "Core libraries that every NodeJS toolchain project should use",
  "main": "lib/index.js",
  "typings": "dist/node-core-library.d.ts",
  "license": "MIT",
  "repository": {
    "url": "https://github.com/microsoft/rushstack.git",
    "type": "git",
    "directory": "libraries/node-core-library"
  },
  "scripts": {
    "build": "heft build --clean",
    "_phase:build": "heft run --only build -- --clean",
    "_phase:test": "heft run --only test -- --clean"
  },
  "dependencies": {
    "fs-extra": "~7.0.1",
    "import-lazy": "~4.0.0",
    "jju": "~1.4.0",
    "resolve": "~1.22.1",
    "semver": "~7.5.4",
    "ajv": "~8.13.0",
    "ajv-draft-04": "~1.0.0",
    "ajv-formats": "~3.0.1"
  },
  "devDependencies": {
    "local-eslint-config": "workspace:*",
    "@rushstack/heft": "0.68.12",
    "@rushstack/heft-node-rig": "2.6.48",
    "@types/fs-extra": "7.0.0",
    "@types/heft-jest": "1.0.1",
    "@types/jju": "1.4.1",
    "@types/node": "18.17.15",
    "@types/resolve": "1.20.2",
    "@types/semver": "7.5.0"
  },
  "peerDependencies": {
    "@types/node": "*"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    }
  }
}
