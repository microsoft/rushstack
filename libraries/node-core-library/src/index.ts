// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Core libraries that every NodeJS toolchain project should use.
 *
 * @packageDocumentation
 */

export { AlreadyReportedError } from './AlreadyReportedError';
export {
  Async,
  AsyncQueue,
  type IAsyncParallelismOptions,
  type IRunWithRetriesOptions,
  type IWeighted
} from './Async';
export type { Brand } from './PrimitiveTypes';
export { FileConstants, FolderConstants } from './Constants';
export { Enum } from './Enum';
export { EnvironmentMap, type IEnvironmentEntry } from './EnvironmentMap';
export {
  type ExecutableStdioStreamMapping,
  type ExecutableStdioMapping,
  type IExecutableResolveOptions,
  type IExecutableSpawnSyncOptions,
  type IExecutableSpawnOptions,
  type IWaitForExitOptions,
  type IWaitForExitWithBufferOptions,
  type IWaitForExitWithStringOptions,
  type IWaitForExitResult,
  type IProcessInfo,
  Executable
} from './Executable';
export { type IFileErrorOptions, type IFileErrorFormattingOptions, FileError } from './FileError';
export type {
  INodePackageJson,
  IPackageJson,
  IPackageJsonDependencyTable,
  IPackageJsonScriptTable,
  IPackageJsonRepository,
  IPeerDependenciesMetaTable,
  IDependenciesMetaTable,
  IPackageJsonExports
} from './IPackageJson';
export {
  Import,
  type IImportResolveOptions,
  type IImportResolveAsyncOptions,
  type IImportResolveModuleOptions,
  type IImportResolveModuleAsyncOptions,
  type IImportResolvePackageOptions,
  type IImportResolvePackageAsyncOptions
} from './Import';
export { InternalError } from './InternalError';
export {
  type JsonObject,
  type JsonNull,
  JsonSyntax,
  type IJsonFileParseOptions,
  type IJsonFileLoadAndValidateOptions,
  type IJsonFileStringifyOptions,
  type IJsonFileSaveOptions,
  JsonFile
} from './JsonFile';
export {
  type IJsonSchemaErrorInfo,
  type IJsonSchemaFromFileOptions,
  type IJsonSchemaFromObjectOptions,
  type IJsonSchemaLoadOptions,
  type IJsonSchemaValidateOptions,
  type IJsonSchemaValidateObjectWithOptions,
  JsonSchema,
  type JsonSchemaVersion
} from './JsonSchema';
export { LockFile } from './LockFile';
export { MapExtensions } from './MapExtensions';
export { MinimumHeap } from './MinimumHeap';
export { PosixModeBits } from './PosixModeBits';
export { ProtectableMap, type IProtectableMapParameters } from './ProtectableMap';
export { type IPackageJsonLookupParameters, PackageJsonLookup } from './PackageJsonLookup';
export {
  PackageName,
  PackageNameParser,
  type IPackageNameParserOptions,
  type IParsedPackageName,
  type IParsedPackageNameOrError
} from './PackageName';
export {
  Path,
  type FileLocationStyle,
  type IPathFormatFileLocationOptions,
  type IPathFormatConciselyOptions
} from './Path';
export { Encoding, Text, NewlineKind, type IReadLinesFromIterableOptions } from './Text';
export { Sort } from './Sort';
export {
  AlreadyExistsBehavior,
  FileSystem,
  type FileSystemCopyFilesAsyncFilter,
  type FileSystemCopyFilesFilter,
  type FolderItem,
  type FileSystemStats,
  type IFileSystemCopyFileBaseOptions,
  type IFileSystemCopyFileOptions,
  type IFileSystemCopyFilesAsyncOptions,
  type IFileSystemCopyFilesOptions,
  type IFileSystemCreateLinkOptions,
  type IFileSystemDeleteFileOptions,
  type IFileSystemMoveOptions,
  type IFileSystemReadFileOptions,
  type IFileSystemReadFolderOptions,
  type IFileSystemUpdateTimeParameters,
  type IFileSystemWriteBinaryFileOptions,
  type IFileSystemWriteFileOptions
} from './FileSystem';
export { FileWriter, type IFileWriterFlags } from './FileWriter';
export { LegacyAdapters, type LegacyCallback } from './LegacyAdapters';
export { StringBuilder, type IStringBuilder } from './StringBuilder';
export { type ISubprocessOptions, SubprocessTerminator } from './SubprocessTerminator';
export { TypeUuid } from './TypeUuid';
