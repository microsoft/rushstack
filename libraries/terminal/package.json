{
  "name": "@rushstack/terminal",
  "version": "0.14.4",
  "description": "User interface primitives for console applications",
  "main": "lib/index.js",
  "typings": "dist/terminal.d.ts",
  "license": "MIT",
  "repository": {
    "url": "https://github.com/microsoft/rushstack.git",
    "type": "git",
    "directory": "libraries/terminal"
  },
  "scripts": {
    "build": "heft build --clean",
    "_phase:build": "heft run --only build -- --clean",
    "_phase:test": "heft run --only test -- --clean"
  },
  "dependencies": {
    "@rushstack/node-core-library": "workspace:*",
    "supports-color": "~8.1.1"
  },
  "devDependencies": {
    "@rushstack/heft": "0.68.12",
    "@rushstack/heft-node-rig": "2.6.48",
    "@types/heft-jest": "1.0.1",
    "@types/node": "18.17.15",
    "@types/supports-color": "8.1.3",
    "local-eslint-config": "workspace:*"
  },
  "peerDependencies": {
    "@types/node": "*"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    }
  }
}
