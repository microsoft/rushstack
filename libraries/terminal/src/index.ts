// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/// <reference types="node" preserve="true" />

/**
 * This library implements a system for processing human readable text that
 * will be output by console applications.
 *
 * @remarks
 * See the {@link TerminalWritable} documentation for an overview of the major concepts.
 *
 * @packageDocumentation
 */

export { type ICallbackWritableOptions, CallbackWritable } from './CallbackWritable';
export { type IDiscardStdoutTransformOptions, DiscardStdoutTransform } from './DiscardStdoutTransform';
export { TerminalChunkKind, type ITerminalChunk } from './ITerminalChunk';
export { MockWritable } from './MockWritable';
export {
  type INormalizeNewlinesTextRewriterOptions,
  NormalizeNewlinesTextRewriter
} from './NormalizeNewlinesTextRewriter';
export { DEFAULT_CONSOLE_WIDTH, PrintUtilities } from './PrintUtilities';
export { RemoveColorsTextRewriter } from './RemoveColorsTextRewriter';
export { type ISplitterTransformOptions, SplitterTransform } from './SplitterTransform';
export { type IStdioLineTransformOptions, StderrLineTransform } from './StdioLineTransform';
export { type IStdioSummarizerOptions, StdioSummarizer } from './StdioSummarizer';
export { StdioWritable } from './StdioWritable';
export { type ITerminalTransformOptions, TerminalTransform } from './TerminalTransform';
export { type ITerminalWritableOptions, TerminalWritable } from './TerminalWritable';
export { type TextRewriterState, TextRewriter } from './TextRewriter';
export { type ITextRewriterTransformOptions, TextRewriterTransform } from './TextRewriterTransform';
export { AnsiEscape, type IAnsiEscapeConvertForTestsOptions } from './AnsiEscape';
export type { ITerminal, TerminalWriteParameters, ITerminalWriteOptions } from './ITerminal';
export { Terminal } from './Terminal';
export { Colorize } from './Colorize';
export { type ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';
export { ConsoleTerminalProvider, type IConsoleTerminalProviderOptions } from './ConsoleTerminalProvider';
export {
  StringBufferTerminalProvider,
  type IStringBufferOutputOptions
} from './StringBufferTerminalProvider';
export {
  PrefixProxyTerminalProvider,
  type IPrefixProxyTerminalProviderOptions,
  type IDynamicPrefixProxyTerminalProviderOptions,
  type IPrefixProxyTerminalProviderOptionsBase,
  type IStaticPrefixProxyTerminalProviderOptions
} from './PrefixProxyTerminalProvider';
export { NoOpTerminalProvider } from './NoOpTerminalProvider';
export { TerminalStreamWritable, type ITerminalStreamWritableOptions } from './TerminalStreamWritable';
