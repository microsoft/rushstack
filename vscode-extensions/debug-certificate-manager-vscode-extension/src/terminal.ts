// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminal } from '@rushstack/terminal';
import * as vscode from 'vscode';
import { stripVTControlCharacters } from 'node:util';

export async function runWorkspaceCommandAsync({
  terminalOptions,
  commandLine,
  terminal
}: {
  terminalOptions: vscode.TerminalOptions;
  commandLine: string;
  terminal: ITerminal;
}): Promise<string> {
  const vsTerminal: vscode.Terminal = vscode.window.createTerminal(terminalOptions);

  // wait for shell to bootup and vs code shell integration to kick-in
  const shellIntegration: vscode.TerminalShellIntegration =
    vsTerminal.shellIntegration ??
    (await new Promise((resolve, reject) => {
      const shellIntegrationDisposable: vscode.Disposable = vscode.window.onDidChangeTerminalShellIntegration(
        (event) => {
          if (event.terminal !== vsTerminal) {
            return;
          }

          resolve(event.shellIntegration);
          shellIntegrationDisposable?.dispose();
        }
      );
    }));

  // Run the command through shell integration and grab output
  return new Promise<string>((resolve, reject) => {
    let outputStream: string = '';

    // start output capturing with the start execution event
    const startExecutionDisposable: vscode.Disposable = vscode.window.onDidStartTerminalShellExecution(
      async (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }
        terminal.writeLine(`Terminal shell execution started`);

        for await (const chunk of event.execution.read()) {
          outputStream += chunk;
        }
      }
    );

    // collect output and exit code
    const endExecutionDisposable: vscode.Disposable = vscode.window.onDidEndTerminalShellExecution(
      (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }

        terminal.writeLine(`Terminal shell execution ended with exit code ${event.exitCode}`);
        outputStream = outputStream.trim();
        outputStream = stripVTControlCharacters(outputStream);
        terminal.writeLine(`Terminal output: ${outputStream}`);

        endExecutionDisposable.dispose();
        startExecutionDisposable.dispose();

        if (event.exitCode === 0) {
          resolve(outputStream);
        } else {
          reject(outputStream);
        }
      }
    );

    shellIntegration.executeCommand(commandLine);
    terminal.writeLine(`Executing command: ${commandLine}`);
  }).finally(() => {
    vsTerminal.dispose();
  });
}
