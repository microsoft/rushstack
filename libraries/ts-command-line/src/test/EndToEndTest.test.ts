// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { ChildProcess } from 'node:child_process';
import { AnsiEscape } from '@rushstack/terminal';
import { Executable } from '@rushstack/node-core-library';

const TEST_CLI_PATH: string = `${__dirname}/test-cli/start`;

function runTestCliTestWithArgs(testName: string, args: string[]): void {
  it(testName, async () => {
    const testCliProcess: ChildProcess = Executable.spawn(process.argv0, [TEST_CLI_PATH, ...args], {
      stdio: 'pipe'
    });
    const { stdout, stderr, exitCode, signal } = await Executable.waitForExitAsync(testCliProcess, {
      encoding: 'utf8'
    });

    expect(stdout).toMatchSnapshot('process stdout');
    expect(stderr).toMatchSnapshot('process stderr');
    expect(exitCode).toMatchSnapshot('process exit code');
    expect(signal).toMatchSnapshot('process signal');
  });
}

describe('end-to-end test', () => {
  beforeEach(() => {
    // ts-command-line calls process.exit() which interferes with Jest
    jest.spyOn(process, 'exit').mockImplementation((code?: number) => {
      throw new Error(`Test code called process.exit(${code})`);
    });
  });

  it(`prints the help`, async () => {
    const { WidgetCommandLine } = await import('./test-cli/WidgetCommandLine');

    const parser = new WidgetCommandLine();

    const globalHelpText: string = AnsiEscape.formatForTests(parser.renderHelpText());
    expect(globalHelpText).toMatchSnapshot('global help');

    for (const action of parser.actions) {
      const actionHelpText: string = AnsiEscape.formatForTests(action.renderHelpText());
      expect(actionHelpText).toMatchSnapshot(action.actionName);
    }
  });

  describe('execution tests', () => {
    runTestCliTestWithArgs('with no args', []);
    runTestCliTestWithArgs('run', ['run']);
    runTestCliTestWithArgs('run --title My Title', ['run', '--title', 'My Title']);
    runTestCliTestWithArgs('run --title My Title --remaining --args', [
      'run',
      '--title',
      'My Title',
      '--remaining',
      '--args'
    ]);
    runTestCliTestWithArgs('push', ['push']);
    runTestCliTestWithArgs('push --force', ['push', '--force']);
    runTestCliTestWithArgs('push --protocol ftp', ['push', '--protocol', 'ftp']);
    runTestCliTestWithArgs('push --protocol bogus', ['push', '--protocol', 'bogus']);
  });
});
