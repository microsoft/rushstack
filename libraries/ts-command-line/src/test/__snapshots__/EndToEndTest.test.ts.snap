// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`end-to-end test execution tests push --force: process exit code 1`] = `0`;

exports[`end-to-end test execution tests push --force: process signal 1`] = `null`;

exports[`end-to-end test execution tests push --force: process stderr 1`] = `""`;

exports[`end-to-end test execution tests push --force: process stdout 1`] = `
"Business logic configured the logger: verbose=false
Received parameters: force=true, protocol=\\"scp\\"
Business logic did the work.
"
`;

exports[`end-to-end test execution tests push --protocol bogus: process exit code 1`] = `2`;

exports[`end-to-end test execution tests push --protocol bogus: process signal 1`] = `null`;

exports[`end-to-end test execution tests push --protocol bogus: process stderr 1`] = `
"widget push: error: argument \\"--protocol\\": Invalid choice: bogus (choose from [ftp, webdav, scp])

"
`;

exports[`end-to-end test execution tests push --protocol bogus: process stdout 1`] = `
"usage: widget push [-h] [-f] [--protocol {ftp,webdav,scp}]
"
`;

exports[`end-to-end test execution tests push --protocol ftp: process exit code 1`] = `0`;

exports[`end-to-end test execution tests push --protocol ftp: process signal 1`] = `null`;

exports[`end-to-end test execution tests push --protocol ftp: process stderr 1`] = `""`;

exports[`end-to-end test execution tests push --protocol ftp: process stdout 1`] = `
"Business logic configured the logger: verbose=false
Received parameters: force=false, protocol=\\"ftp\\"
Business logic did the work.
"
`;

exports[`end-to-end test execution tests push: process exit code 1`] = `0`;

exports[`end-to-end test execution tests push: process signal 1`] = `null`;

exports[`end-to-end test execution tests push: process stderr 1`] = `""`;

exports[`end-to-end test execution tests push: process stdout 1`] = `
"Business logic configured the logger: verbose=false
Received parameters: force=false, protocol=\\"scp\\"
Business logic did the work.
"
`;

exports[`end-to-end test execution tests run --title My Title --remaining --args: process exit code 1`] = `2`;

exports[`end-to-end test execution tests run --title My Title --remaining --args: process signal 1`] = `null`;

exports[`end-to-end test execution tests run --title My Title --remaining --args: process stderr 1`] = `
"widget run: error: Unrecognized arguments: --remaining --args.

"
`;

exports[`end-to-end test execution tests run --title My Title --remaining --args: process stdout 1`] = `
"usage: widget run [-h] [--title TITLE] ...
"
`;

exports[`end-to-end test execution tests run --title My Title: process exit code 1`] = `0`;

exports[`end-to-end test execution tests run --title My Title: process signal 1`] = `null`;

exports[`end-to-end test execution tests run --title My Title: process stderr 1`] = `""`;

exports[`end-to-end test execution tests run --title My Title: process stdout 1`] = `
"Business logic configured the logger: verbose=false
Console Title: My Title
Arguments to be executed: []
"
`;

exports[`end-to-end test execution tests run: process exit code 1`] = `0`;

exports[`end-to-end test execution tests run: process signal 1`] = `null`;

exports[`end-to-end test execution tests run: process stderr 1`] = `""`;

exports[`end-to-end test execution tests run: process stdout 1`] = `
"Business logic configured the logger: verbose=false
Console Title: (none)
Arguments to be executed: []
"
`;

exports[`end-to-end test execution tests with no args: process exit code 1`] = `0`;

exports[`end-to-end test execution tests with no args: process signal 1`] = `null`;

exports[`end-to-end test execution tests with no args: process stderr 1`] = `""`;

exports[`end-to-end test execution tests with no args: process stdout 1`] = `
"usage: widget [-h] [-v] <command> ...

The \\"widget\\" tool is a code sample for using the @rushstack/ts-command-line 
library.

Positional arguments:
  <command>
    push         Pushes a widget to the service
    run          This action (hypothetically) passes its command line 
                 arguments to the shell to be executed.

Optional arguments:
  -h, --help     Show this help message and exit.
  -v, --verbose  Show extra logging detail

[1mFor detailed help about a specific command, use: widget <command> -h[22m
"
`;

exports[`end-to-end test prints the help: global help 1`] = `
"usage: widget [-h] [-v] <command> ...

The \\"widget\\" tool is a code sample for using the @rushstack/ts-command-line 
library.

Positional arguments:
  <command>
    push         Pushes a widget to the service
    run          This action (hypothetically) passes its command line 
                 arguments to the shell to be executed.

Optional arguments:
  -h, --help     Show this help message and exit.
  -v, --verbose  Show extra logging detail

[bold]For detailed help about a specific command, use: widget <command> -h[normal]
"
`;

exports[`end-to-end test prints the help: push 1`] = `
"usage: widget push [-h] [-f] [--protocol {ftp,webdav,scp}]

Here we provide a longer description of how our action works.

Optional arguments:
  -h, --help            Show this help message and exit.
  -f, --force           Push and overwrite any existing state
  --protocol {ftp,webdav,scp}
                        Specify the protocol to use. This parameter may 
                        alternatively be specified via the WIDGET_PROTOCOL 
                        environment variable. The default value is \\"scp\\".
"
`;

exports[`end-to-end test prints the help: run 1`] = `
"usage: widget run [-h] [--title TITLE] ...

This demonstrates how to use the defineCommandLineRemainder() API.

Positional arguments:
  \\"...\\"          The remaining arguments are passed along to the command 
                 shell.

Optional arguments:
  -h, --help     Show this help message and exit.
  --title TITLE  An optional title to show in the console window. This 
                 parameter may alternatively be specified via the 
                 WIDGET_TITLE environment variable.
"
`;
